### PR TITLE
Feat/proxy v0.1

### PR DESCRIPTION
**Pull request by m4ntis**
#### Hey all,

As some of you maintainers probably know by now, for the last couple of months me and 2 colleagues of mine (@kubistika and @speidy) have been working on implementing an RDP proxy using the library. This is a PR to introduce what we've been working on up till now, containing a working base version of the proxy.

The changes and number of commits are quite vast, so I'll start with the motivation, why we chose to implement it in this way and why the *h* you might want all of this massive PR. Later on I'll get to what is implemented here, how to use it, future goals and where we want to get to with this proxy.

## Motivation
The proxy came about for a number of reasons, mostly security related:
- There is no real good RDP proxy out there, enabling you to finely select the specific features and channels you want to allow, filter or right out disable for all connections passing through. Lets say you have some isolated network where you want to minimize all traffic going in and out while still enabling easy access via RDP, you probably don't want any channels passing through but the bare minimum, say GFX from the server and basic input events from the client.
- There is no proxy that lets you monitor and analyze RDP traffic going through it. We might want to log certain events or look for anomalies, such as suspicious amounts of certain events or specific channel messages.
- We wanted a proxy that will let you capture and record all traffic passing through. You might use this feature by placing the proxy as the only gateway to some sensitive server such as a DC, being able to view all passed connections filtered by time and username.
- We wanted to have a proxy that has an API for extra authentication, letting us implement a 2FA towards the proxy and disallowing connections to the target server without authentication.

## Why we chose FreeRDP
So the reasoning for choosing to implement the proxy as a part of the FreeRDP library is quite obvious. FreeRDP is by far the most maintained and up to date OS implementation of the protocol, and with the server side implementations only getting bigger, it only seemed fitting. As we **do** parse the messages going through the proxy (the whole point of the proxy), in order to support updated versions of the protocol we need to use an updating implementation of RDP. There is a tiny, basic proxy MITM in [pyrdp](https://github.com/GoSecure/pyrdp) but it's too small, not extensive and with way lower chances of being properly maintained in the long run.

## Benefits to the library
Well first of all, we hope that the proxy in itself is a useful thing to have on it's own. But further than that, as you've seen in the last month or so, me and @kubistika already issued a few small PRs that fixed some of the server side implementation of the protocol. This is just the beginning, and this PR already has a few other fixes to GFX parsing bundled inside it (I'll elaborate below). The proxy tests both client and server side implementations of the protocol without having to implement a client nor server, as you can easily use mstsc and Microsoft's RDP TS server going talking via the proxy to quickly find any parsing errors.

I'm sure we'll continue to squish bugs as we continue to other parts of the protocol :wink:.

Additionally, the proxy is an incentive for us to implement the server side of a few other channels that aren't implemented yet, as we want to parse and proxy them too (I am currently working on a RAIL server implementation, might be a few other coming after that).

## So what do we have here
This is the first usable version of the proxy, so most of our goals aren't yet fully implemented (some not at all). What we do have is a proxy that can parse and proxy connections, passing graphics messages and keyboard / mouse events. GFX and bitmap updates are fully working, GDI still has some issues but I think that it's obvious why GDI is a low priority right now.

**Basic list of features**:
- GFX, bitmap updates and input events parsing and proxying.
- Settings sync: A client connects to and negociates with the proxy, the proxy takes the negociated settings and connects with them to the server, and then renegociates with the client using the final settings.
- GFX capabilities sync: The client and server capabilities are properly synced.
- Basic configuration (example below). Not all features in the config file are fully implemented yet, but soon to be.

There are a few changes to the library itself too:
#### Bugfixes:
1. In `client/rdpgfx_main.c:rdpgfx_recv_caps_confirm_pdu()`, the caps data length was incorrectly read into a local variable.
2. In `client/rdpgfx_main.c:rdpgfx_recv_map_surface_to_scaled_window_pdu()` and `scaled_output_pdu`, the stream length magic number was incorrect, as the methods were probably copy+pasted from `recv_map_surface_to_window/output` without changing the length.
3. Same issue as the one above in the server side, in `server/rdpgfx_main.c:rdpgfx_send_map_surface_to_scaled_window_pdu`, stream length is incorrect.
4. Same again with `scaled_output_pdu` on the server side, but on this PDU the values in the struct were named wrongly by the spec, `targetX` and `targetY` instead of `targetWidth` and `targetHeight`.

#### Changes to API:
1. In order to properly be able to forward GFX frame acking and to sync capabilities between the client and server, we needed to have a way to have control over sending FrameAck and CapsConfirm messages, so we exposed these methods as part of the `RdpgfxClientContext`'s API.
2. Additionally, we needed some way of disabling the client's automatic frame acking / caps confirmation, without hindering default functionality. We added and optional `OnOpen` callback with flags that control whether to `do_frame_acks` and `do_caps_confirm`. These values default to `TRUE`.

## Usage
To compile with cmake, the `-DWITH_SERVER=on` flag will compile the proxy. Running it is as simple as `./pfreerdp` (there are no CLI options yet). There is a `config.ini` present in `server/proxy` that currently has to be in the working dir of the proxy. Lets go over the config:

```ini
[Server]
Host = "0.0.0.0"
Port = 3389
LocalOnly = 0

[Target]
UseLoadBalanceInfo = 1
Host = "Target server"
Port = 3389

[Input]
Mouse = 1
Keyboard = 1

[Graphics]
GFX = 1
BitmapUpdate = 1

[Security]
NlaSupport = 0
TlsSupport = 0
RdpSupport = 0

[Channels]
WhitelistMode = 0
AllowedChannels = "cliprdr,Microsoft::Windows::RDS::Video::Control"
DeniedChannels = "Microsoft::Windows::RDS::Geometry"
```
All sections but the `[Target]` and `[Channels]` are quite obvious (host config and allowing / disallowing RDP features).

The `[Target]` config option regards the method of conveying which server shall the client connecting to the proxy be forwarded to. The simple option is configuring a `Host` and `Port` value and setting `UseLoadBalanceInfo` to `0`. This fixes all connections to a single configured destination. The second method is to dynamically detect the target for each incoming connection. The way we figured out is to use the routing token to send that info (this is one of the things we would like to hear your opinion on). This allows the proxy to receive the target server info from any normal rdp client. For mstsc, you could use the `loadbalanceinfo:s:` option in an rdp file, and in any of the FreeRDP clients you could use the `/load-balance-info:` flag. The format of the routing token is `Cookie: msts=<token>`, so as an example, using with xfreerdp would look like this:

`xfreerdp /v:<proxy addr> /load-balance-info:"Cookie: msts=<destination server>"`
 
The `[Channels]` section is not yet implemented, but expresses the two modes we want the proxy to be able to work in, blacklist and whitelist. The first would allow all channels excluding the ones configured in `DeniedChannels`. The latter would allow only the channels configured in `AllowedChannels`. Currently, only the channels explicitly supported by the proxy are allowed. In order to support a blacklist mode, we would have to find a way of proxing channels without inspecting the data. We thought about implementing a "dummy" channel that just exposes receiving and sending data (we'd love hearing suggestions about this too).

### Small caveat
NLA isn't supported yet, but will be very soon.

## Looking forward
I'll keep this short, but here's a short summary of the features we want to implement:

1. A hooking type of API for events in the connection, probably loading "so" files that register hooks the proxy exposes.
2. Implementing more channels.
3. Implementing a "blacklist" channel mode (explained just above)
4. Adding some sort of API for a third party authentication logic.
5. Connection recording functionality.

This list can change / grow, we're willing to hear suggestions for features / how to implement them.

Please note that this is just the first initial PR for this proxy. We are and will be actively developing this proxy, adding support for more channels and features. We wanted to do this PR and merge early so that we could get a lot of feedback and suggestions from the maintainers, and so that you follow and review our work.

Thanks for taking the time to read and review this big PR :)

### CODE REVIEW

File: include/freerdp/server/rdpgfx.h [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #define FREERDP_CHANNEL_RDPGFX_SERVER_RDPGFX_H
 
 #include <freerdp/channels/rdpgfx.h>
+#include <freerdp/freerdp.h>
```
Comment (speidy) [inline include/freerdp/server/rdpgfx.h]: shouldn't be here , but in pf_server.c i think

File: server/proxy/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  pf_common.h
+  pf_gdi.c
+  pf_gdi.h
+  pf_config.c
+  pf_config.h
+  pf_graphics.c
+  pf_graphics.h
+  pf_log.h)
+
+	# On windows create dll version information.
```
Comment (speidy) [inline server/proxy/CMakeLists.txt]: take me back

File: server/proxy/pf_channels.c [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+#include "pf_log.h"
+
+#define TAG PROXY_TAG("channels")
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT pf_encomsp_participant_created(EncomspClientContext* context,
```
Comment (speidy) [inline server/proxy/pf_channels.c]: maybe remove `encomsp` stuff for now

File: server/proxy/pf_channels.c [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	pServerContext* ps = pc->pdata->ps;
+	RdpgfxClientContext* gfx;
+	RdpgfxServerContext* server;
+	WLog_DBG(TAG, "Channel connected: %s", e->name);
+
+	if (strcmp(e->name, RDPEI_DVC_CHANNEL_NAME) == 0)
+	{
+		pc->rdpei = (RdpeiClientContext*) e->pInterface;
+	}
+	else if (strcmp(e->name, TSMF_DVC_CHANNEL_NAME) == 0)
```
Comment (speidy) [inline server/proxy/pf_channels.c]: rm unhandled channels ?

File: server/proxy/pf_channels.c [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	else if (strcmp(e->name, RAIL_SVC_CHANNEL_NAME) == 0)
+	{
+	}
+	else if (strcmp(e->name, CLIPRDR_SVC_CHANNEL_NAME) == 0)
+	{
+	}
+	else if (strcmp(e->name, ENCOMSP_SVC_CHANNEL_NAME) == 0)
+	{
+		pf_encomsp_init(pc, (EncomspClientContext*) e->pInterface);
```
Comment (speidy) [inline server/proxy/pf_channels.c]: whats this `encomsp` ?? :|

File: server/proxy/pf_client.c [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+#define TAG PROXY_TAG("client")
+
+/**
+ * Re-negociate with original client after negociation between the proxy
+ * and the target has finished.
+ */
+void proxy_server_reactivate(rdpContext* client, rdpContext* target)
+{
+	pf_common_copy_settings(client->settings, target->settings);
+	client->update->DesktopResize(client);
```
Comment (speidy) [inline server/proxy/pf_client.c:65]: @kubistika  add comment why we call this func

File: server/proxy/pf_client.c [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+		offscreen_cache_register_callbacks(instance->update);
+		palette_cache_register_callbacks(instance->update);
+	}
+
+	update->BeginPaint = pf_client_begin_paint;
+	update->EndPaint = pf_client_end_paint;
+	update->BitmapUpdate = pf_client_bitmap_update;
+	update->DesktopResize = pf_client_desktop_resize;
+	rdpContext* ps = (rdpContext*) pc->pdata->ps;
+	proxy_server_reactivate(ps, context);
```
Comment (speidy) [inline server/proxy/pf_client.c:199]: should we call this here?

File: server/proxy/pf_client.c [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+	instance->PostConnect = pf_client_post_connect;
+	instance->PostDisconnect = pf_client_post_disconnect;
+	instance->VerifyCertificateEx = pf_client_verify_certificate_ex;
+	instance->VerifyChangedCertificateEx = pf_client_verify_changed_certificate_ex;
+	instance->LogonErrorInfo = pf_logon_error_info;
+	return TRUE;
+}
+
+
+static void pf_client_client_free(freerdp* instance, rdpContext* context) {}
```
Comment (speidy) [inline server/proxy/pf_client.c]: nothing to do here?

File: server/proxy/pf_common.c [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	/* GFX */
+	dst->GfxThinClient = src->GfxThinClient;
+	dst->GfxSmallCache = src->GfxSmallCache;
+	dst->GfxProgressive = src->GfxProgressive;
+	dst->GfxProgressiveV2 = src->GfxProgressiveV2;
+	dst->GfxH264 = src->GfxH264;
+	dst->GfxAVC444 = src->GfxAVC444;
+	dst->GfxSendQoeAck = src->GfxSendQoeAck;
+	dst->GfxAVC444v2 = src->GfxAVC444v2;
+}
```
Comment (speidy) [inline server/proxy/pf_common.c:107]: new line @ eof

File: server/proxy/pf_context.c [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+BOOL client_to_proxy_context_new(freerdp_peer* client,
+                                 pServerContext* context)
+{
+	context->vcm = WTSOpenServerA((LPSTR) client->context);
+
+	if (!context->vcm || context->vcm == INVALID_HANDLE_VALUE)
+		goto fail_open_server;
+
+	return TRUE;
+fail_open_server:
```
Comment (speidy) [inline server/proxy/pf_context.c:36]: add new line before label

File: server/proxy/pf_rdpgfx.c [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+	gfx->DeleteSurface = pf_rdpgfx_delete_surface;
+	gfx->SolidFill = pf_rdpgfx_solid_fill;
+	gfx->SurfaceToSurface = pf_rdpgfx_surface_to_surface;
+	gfx->SurfaceToCache = pf_rdpgfx_surface_to_cache;
+	gfx->CacheToSurface = pf_rdpgfx_cache_to_surface;
+	gfx->CacheImportReply = pf_rdpgfx_cache_import_reply;
+	gfx->EvictCacheEntry = pf_rdpgfx_evict_cache_entry;
+	gfx->MapSurfaceToOutput = pf_rdpgfx_map_surface_to_output;
+	gfx->MapSurfaceToWindow = pf_rdpgfx_map_surface_to_window;
+    gfx->MapSurfaceToScaledOutput = pf_rdpgfx_map_surface_to_scaled_output;
```
Comment (speidy) [inline server/proxy/pf_rdpgfx.c]: spaces to tabs

File: server/proxy/pf_server.c [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+#include "pf_log.h"
+#include "pf_config.h"
+#include "pf_client.h"
+#include "pf_context.h"
+#include "pf_input.h"
+#include "pf_update.h"
+
+#define TAG PROXY_TAG("server")
+
+BOOL pf_server_rdpgfx_init(pServerContext* ps)
```
Comment (speidy) [inline server/proxy/pf_server.c]: move to `pf_rdpgfx.c`

File: server/proxy/pf_server.c [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+			return FALSE;
+		}
+
+		WLog_DBG(TAG, "Parsed target from load-balance-info: %s:%i", host, port);
+	}
+	else
+	{
+		/* use hardcoded target info from configuration */
+		host = _strdup(config->TargetHost);
+		port = config->TargetPort > 0 ? config->TargetPort : port;
```
Comment (speidy) [inline server/proxy/pf_server.c:156]: `&& config->TargetPort <= USHORT_MAX`

File: server/proxy/pf_server.c [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+		WLog_DBG(TAG, "Using hardcoded target host: %s:%i", host, port);
+	}
+
+	/* Start a proxy's client in it's own thread */
+	pc = (pClientContext*) p_client_context_create(client->settings, host, port);
+	pc->pdata = ps->pdata;
+	/* Inject proxy's client context to proxy's context */
```
Comment (speidy) [inline server/proxy/pf_server.c]: Inject -> keep

File: server/proxy/pf_server.c [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+	{
+		freerdp_peer_free(client);
+		return 0;
+	}
+
+	ps = (pServerContext*) client->context;
+	ps->dynvcReady = CreateEvent(NULL, TRUE, FALSE, NULL);
+	pdata = calloc(1, sizeof(proxyData));
+	ps->pdata = pdata;
+	/* inject configuration to proxyData */
```
Comment (speidy) [inline server/proxy/pf_server.c]: again , not inject

File: server/proxy/pf_server.c [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+	pdata->config = client->ContextExtra;
+	client->settings->SupportGraphicsPipeline = TRUE;
+	client->settings->SupportDynamicChannels = TRUE;
+	/* TODO: Read path from config and default to /etc */
+	client->settings->CertificateFile = _strdup("server.crt");
```
Comment (speidy) [inline server/proxy/pf_server.c:228]: lets add to `config.ini`, its easy

File: server/proxy/pf_server.c [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+	client->settings->PrivateKeyFile = _strdup("server.key");
+	client->settings->RdpKeyFile = _strdup("server.key");
```
Comment (speidy) [inline server/proxy/pf_server.c:230]: rm, its not relevant

File: server/proxy/pf_server.c [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+
+	if (!client->settings->CertificateFile || !client->settings->PrivateKeyFile
+	    || !client->settings->RdpKeyFile)
+	{
+		WLog_ERR(TAG, "Memory allocation failed (strdup)");
```
Comment (speidy) [inline server/proxy/pf_server.c:235]: change the log so it mentions `CertificateFile`  `PrivateKeyFile` etc..

File: include/freerdp/server/rdpgfx.h (hunk 1 above)
Comment (m4ntis) [inline include/freerdp/server/rdpgfx.h]: Need to ensure this is legit

File: server/proxy/pf_server.c (hunk 18 above)
Comment (speidy) [inline server/proxy/pf_server.c:235]: change the log so it mentions `CertificateFile`  `PrivateKeyFile` etc..

File: libfreerdp/core/nego.c [hunk 19]
```diff
@@ -36,34 +36,6 @@
 
 #define TAG FREERDP_TAG("core.nego")
 
-struct rdp_nego
```
Comment (m4ntis) [inline libfreerdp/core/nego.c]: @kubistika I can't remember why we did this

File: server/proxy/pf_server.c [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+
+				break;
+
+			default:
+				break;
+		}
+	}
+
+fail:
+
```
Comment (speidy) [inline server/proxy/pf_server.c:327]: rm blank line

**Review by speidy — CHANGES_REQUESTED** [review]
Take a look at my comments, lets merge

File: libfreerdp/core/nego.c (hunk 19 above)
Comment (speidy) [reply libfreerdp/core/nego.c]: i did that so we can get the routing token from our code.

File: include/freerdp/server/rdpgfx.h (hunk 1 above)
Comment (speidy) [reply include/freerdp/server/rdpgfx.h]: it wasn't there, and doesn't make sense

**Comment by speidy** [comment]
we should squash the `fixup!` commits using `git rebase -i --autosquash origin/master`

File: include/freerdp/server/rdpgfx.h (hunk 1 above)
Comment (m4ntis) [reply include/freerdp/server/rdpgfx.h]: I'm not sure it compiles without it tho, can you please test that?

File: libfreerdp/core/nego.c (hunk 19 above)
Comment (m4ntis) [reply libfreerdp/core/nego.c]: So can we revert it

File: include/freerdp/server/rdpgfx.h (hunk 1 above)
Comment (m4ntis) [reply include/freerdp/server/rdpgfx.h]: test it

File: server/proxy/pf_channels.c (hunk 3 above)
Comment (m4ntis) [reply server/proxy/pf_channels.c]: oh yeah for sure, there is the encomsp in pf_context.h too if i remember correctly, remove it all

File: server/proxy/pf_channels.c (hunk 4 above)
Comment (m4ntis) [reply server/proxy/pf_channels.c]: yes

File: server/proxy/pf_channels.c (hunk 5 above)
Comment (m4ntis) [reply server/proxy/pf_channels.c]: i answered above

File: server/proxy/pf_server.c (hunk 16 above)
Comment (m4ntis) [reply server/proxy/pf_server.c:228]: do it

File: server/proxy/pf_server.c (hunk 17 above)
Comment (m4ntis) [reply server/proxy/pf_server.c:230]: do it

File: server/proxy/pf_server.c (hunk 18 above)
Comment (m4ntis) [reply server/proxy/pf_server.c:235]: go for it

File: server/proxy/pf_server.c (hunk 20 above)
Comment (m4ntis) [reply server/proxy/pf_server.c:327]: yes

File: server/proxy/pf_client.c (hunk 7 above)
Comment (kubistika) [reply server/proxy/pf_client.c:199]: What do you mean by this? The reactivation? Yes, we reactivate right after the connection

File: server/proxy/pf_server.c (hunk 20 above)
Comment (kubistika) [reply server/proxy/pf_server.c:327]: but the formatter adds it

File: server/proxy/pf_server.c (hunk 17 above)
Comment (kubistika) [reply server/proxy/pf_server.c:230]: done

File: server/proxy/CMakeLists.txt (hunk 2 above)
Comment (kubistika) [reply server/proxy/CMakeLists.txt]: done

File: libfreerdp/core/nego.c (hunk 19 above)
Comment (kubistika) [reply libfreerdp/core/nego.c]: No, we use rdpNego struct in pf_server to get the routing token and extract target host information. This struct should remain exported.

**Comment by speidy** [comment]
lets close this one

**Pull request closed without merging**